text/3424-cargo-script.md

```diff
@@ -0,0 +1,593 @@
+- Feature Name: cargo-script
+- Start Date: 2023-04-26
+- Pre-RFC: [internals](https://internals.rust-lang.org/t/pre-rfc-cargo-script-for-everyone/18639)
+- eRFC PR: [rust-lang/rfcs#3424](https://github.com/rust-lang/rfcs/pull/3424)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This *experimental RFC* adds unstable support for single-file
+packages in cargo so we can explore the design and resolve
+questions with an implementation to collect feedback on.
+
+Single-file packages are `.rs` files with an embedded
+manifest.  These will be accepted with just like `Cargo.toml` files with
+`--manifest-path`.  `cargo` will be modified to accept `cargo <file>.rs` as a
+shortcut to `cargo run --manifest-path <file>.rs`.  This allows placing
+`cargo` in a `#!` line for directly running these files.
+
+Example:
+```rust
+#!/usr/bin/env cargo
+
+//! ```cargo
+//! [dependencies]
+//! clap = { version = "4.2", features = ["derive"] }
+//! ```
```

Comment:
Does this need to be a doc comment? Could it just be an ordinary comment? If so, I think it should be, to make it more obvious that this isn't magic syntax.

```suggestion
// ```cargo
// [dependencies]
// clap = { version = "4.2", features = ["derive"] }
// ```
```

Reply:
The reference-level explanation should explain that this can be a regular `//` comment, a `///` doc comment, or a `//!` file-level doc comment.

Reply:
The strawman examples in the eRFC are using doc-comments.  I've called out how we store the embedded manifest as one of the unresolved questions that we plan to resolve before stablization.  During the Pre-RFC, we [explored a lot of options](https://github.com/epage/cargo-script-mvs/blob/main/0000-cargo-script.md#embedded-manifest-format).  I left these out of the eRFC to avoid getting side tracked by things that aren't meant to be resolved at this time.

Reply:
@epage We don't have to specify all of that now; I'd just like to make sure that people don't assume this *has* to be a doc comment.

Reply:
The current proposal is for it to be doc comments (an alternative was to instead accept regular comments). @Muscraft is prototyping the initial implementation of this proposal and has switched to `syn` to extract the doc comments, finding that our regex parser for the doc comments was accepting cases that aren't turned into `#![doc]`s.

I've clarified this within the guide section.

Reply:
Wouldn't it be better to have specific syntax, or a macro, instead?

Whenever languages have comments which are parsed and have effects, it feels a bit uneasy. Things which are commented out should never have any effect, I think.

But maybe this is different, since I suppose rustc really does ignore these comments.

Reply:
For more of an idea of what syntaxes we'll consider, see the Pre-RFC. We decided to do an eRFC, rather than a full RFC, and only included a rough sketch to give people an idea of what the experiment will look like, deferring actual decisions to the experiment and approval of the experiment's result.